[Android] Infrastructure for reflexive/responsive settings UI

jahorton opened the pull request:
As noted on #1782, there presently is no auto-updating code for the various elements of the new settings UI.  To facilitate this, this PR introduces some custom `Adapter` types.  Since they're designed to notify `ListView`s and other UI elements of their mutations, `Adapters` are a perfect solution for UI auto-updates.

Of course, there are some inconvenient limitations in the provided implementations.  Toward that end, this PR introduces `NestedAdapters` to allow one central data source to be filtered (useful for language-specific displays, like the one missing for the specific-language settings screen) and reused across different UI `Activity`s.

Due to the hierarchy of the settings UI, I felt it best to start at the deepest point of the UI and work backward.  As such, the first part to be addressed as a practical implementation (and test bed) of the new `Adapter` system is the lexical model installation screen.

For future PRs:
- Implement these changes for the other settings `Activity` windows.
  - These are currently built within `onCreate` with static datasets that would need recomputation. This holds especially true for the language-specific keyboard screen, which is computed in its parent `Activity`!  But that's what these `Adapter`s are for.

keyman-server commented:
[OSK renders for Windows](https://build.palaso.org/repository/download/KeymanDesktop_TestPrRenderOnScreenKeyboards/153414:id/kvk/index.html)

jahorton commented:
In the process of doing followup work, I've noted a few design issues with the code as originally implemented.  The most egregious issue is that the implementation of the `Dataset` type currently requires O(n^2) to build due to lack of optimization.  (As a result, loading the keyboard catalog with n near 3200 keyboards takes forever.)

As a result, I've temporarily marked this as WIP until this is properly addressed.

jahorton commented:
OK, largely worked around that core issue.  Now the main blocking thing is the technically-not-supported use of a general lexical-model query instead of using language-specific ones.

jahorton commented:
OK, finally got the lexical model querying stuff redone.  I've added logic to resend a query for lexical models (invalidating their cache) whenever a new keyboard is added, since that's likely to add a new language code to track.  It's not needed with the placeholder query, but that query is designed to be swapped with a language-specific one once it's implemented.

Queries are also done "as needed" between the keyboards and lexical models.  If lexical models are the only cache invalidated, they'll be the only cache reloaded.  Took a while to get that logic vetted, but that should help things.

I also fixed a couple of bugs with keyboard deletion from the settings menu while I was at it; there were some 'interesting' interactions with the old keyboard picker due to how keyboards are currently maanged within the system, which hasn't yet been replaced.

jahorton commented:
> Testing notes:
> 
>     5. In the list of models for a language, I was able to touch the MTNT model and accept the option to download it. However, it does not update the check mark after doing so. Touching the model again shows the download query again. The model does not appear to be successfully installed? Note, after going out of settings and coming back in, MTNT now shows in the Language Settings view but is still not checked in the Model list for the language.

Interesting; I thought I'd fixed that.  I'll have to look into it.

>     6. After uninstalling the English Example template model, it still showed as selected in the Language Settings view.

Probably related; I figured that one's a lower priority than some of the other outstanding issues, like...

>     8. From the Language Settings view, touching the `(+)` button to add another keyboard for the language shows the list of all languages, not just the list of available keyboards for the language.
> 
>     9. Conceptually, "Add New Keyboard" should probably say "Add New Language" when touching the `(+)` button from the "Languages Settings" view
>
>
> Outside the settings UI:
> 
>     1. The globe language switcher should no longer check for updates (which in the future should be a background task ...)
> 
>     2. The globe language switcher should not have a `(+)` button -- this should only be accessible from Settings.

Like all of those, which will be addressed by a dependent PR and not this one.

Some of the other things are simply what I was given when taking over Android (3, 4, 7, 9) and probably should have been caught before that; I'll aim to address them after getting the more substantial issues handled.  That said...

>     2. Adding a model should not query the keyboard catalog, just the LM catalog.
>
> Outside the settings UI:
> 
>     1. The globe language switcher should no longer check for updates (which in the future should be a background task ...)

Given the 'outside the settings UI' point, I assume that the update check should instead be under "Languages Settings" / "Installed Languages", right?  That update check (when implemented in a dependent PR) will need to check both the keyboard catalog and the lexical models anyway.  The current check is in a subview that would receive the data already queried from the update-checking superview, so this'll naturally change as more of the functionality is properly implemented.

jahorton commented:
At this point, the only items listed above that I expect to remain an issue:

> 3. The different sized views for different parts of settings are a little noisy. I would make them all full-screen, just like the initial settings view.

I've figured out how to start implementing this; #1836 will have this set for the "Installed Language" and "<Language> Settings" screens, though not for the model picker.  It still won't perfectly match the style of the overall settings menu though - that'll take a bit more work.

> 11. The "Show banner" option can be switched off even when predictive text is enabled.

That one is distinct enough to be its own PR; there's already far more than enough content within each PR in the chain to review, after all, and the logic there is comparatively unrelated.

While the longpress crash is handled by #1844, I expect the other PRs in this chain (#1836, #1842) to handle the rest.

Another item for an extra PR, though:  I don't believe there's an in-app setting for the 'preferred' lexical model for a language yet.  Since we have two models for English right now (simple wordlist vs MTNT data), it's something I've seen during testing.

jahorton commented on android/KMEA/app/src/main/java/com/tavultesoft/kmea/ModelPickerActivity.java:
```diff
@@ -150,349 +200,54 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
 
-  // Each language ID will save to a separate cache
-  private static final String cacheFilename(String languageID) {
-    final String jsonCacheFilename = "jsonLexicalModelsCache";
```
Note from here on:  
All async/download/query related functionality in this file was moved to the new `CloudRepository` class to serve as a centralized downloader/updater class.  This code was near-duplicated with that of the `LanguageListActivity` activity's code, and a considerable part of it could be considered near-duplicated with that of the `KeyboardPickerActivity`'s update-check code.

jahorton commented on android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/Dataset.java:
```diff
@@ -0,0 +1,302 @@
+
+public class Dataset extends ArrayAdapter<Dataset.LanguageDataset> {
+  public abstract class LanguageFilter<Type extends LanguageCoded, Adapter extends LanguageCodedAdapter<Type>> implements AdapterFilter<Type, Adapter, String> {
+    abstract Set<Type> getSetFrom(LanguageDataset metadata);
```
Sadly, this class was somewhat prone to iterative development across the PR chain, but this is where the class was originally introduced.

Rather than rely on massive Lists of HashMaps for everything, and then building weird frankenstein-hashmaps for the LanguagesListActivity, this class and its nested classes will provide a cleaner, better-documented, and far more UI-friendly way to manage Keyman resource data.  As it and many of its nested classes are `Adapter`-based, they're designed to pair perfectly with the new `NestedAdapter` class to facilitate reflexive/reactive UI development that provides a good Settings management UX.

The main issue for reviewing is that I missed a few details here that were needed by later work, hence fleshed out in the later PRs.

jahorton commented on android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/adapters/NestedAdapter.java:
```diff
@@ -0,0 +1,179 @@
+/**
+ * An ArrayAdapter subclass that uses another ArrayAdapter (or subclass) as its data source,
+ * propagating relevant changes across both to maintain a 'linked'/'nested' state.
+ * Also permits use of an AdapterFilter to allow using a filtered view of that source.
```
When paired with the `Dataset` class and its nested classes, this new `NestedAdapter` is designed to allow re-use of a single common datastructure across all of the settings UI menus - even if that base data structure must be accessed by semi-global / `static` singleton on the `CloudRepository` class (for cloud stuff) and from the `KeyboardPickerActivity` (for installed resources).

This class then serves as the heart of the event-messaging functionality needed to drive our desired UX in the Settings menus.
